orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java

```diff
@@ -20,50 +20,271 @@
+          .map(maxStats ->
+              (int) ChronoUnit.DAYS.between(EPOCH_DAY,
+                  EPOCH.plus(maxStats.getTime(), ChronoUnit.MILLIS).toLocalDate()))
+          .orElse(null);
+    } else if (columnStats instanceof TimestampColumnStatistics) {
+      TimestampColumnStatistics tColStats = (TimestampColumnStatistics) columnStats;
+      Timestamp maxValue = tColStats.getMaximumUTC();
+      max = Optional.ofNullable(maxValue)
+          .map(v -> TimeUnit.MILLISECONDS.toMicros(v.getTime()))
+          .map(v -> v + 1_000) // Add 1 millisecond to handle precision issue due to ORC-611
+          .orElse(null);
+    } else if (columnStats instanceof BooleanColumnStatistics) {
+      BooleanColumnStatistics booleanStats = (BooleanColumnStatistics) columnStats;
+      max = booleanStats.getTrueCount() > 0;
+    }
+    return Optional.ofNullable(Conversions.toByteBuffer(column.type(), max));
+  }
+
+  private static Set<TypeDescription> findColumnsInContainers(Schema schema,
+                                                              TypeDescription orcSchema) {
+    ColumnsInContainersVisitor visitor = new ColumnsInContainersVisitor();
+    OrcSchemaWithTypeVisitor.visit(schema, orcSchema, visitor);
+    return visitor.getColumnsInContainers();
+  }
+
+  private static class ColumnsInContainersVisitor extends OrcSchemaWithTypeVisitor<TypeDescription> {
+
+    private final Set<TypeDescription> columnsInContainers;
+
+    private ColumnsInContainersVisitor() {
+      columnsInContainers = Sets.newHashSet();
+    }
+
+    public Set<TypeDescription> getColumnsInContainers() {
+      return columnsInContainers;
+    }
+
+    private Set<TypeDescription> flatten(TypeDescription rootType) {
+      if (rootType == null) {
+        return ImmutableSet.of();
+      }
+
+      final Set<TypeDescription> flatTypes = Sets.newHashSetWithExpectedSize(rootType.getMaximumId());
+      final Queue<TypeDescription> queue = Queues.newLinkedBlockingQueue();
+      queue.add(rootType);
+      while (!queue.isEmpty()) {
+        TypeDescription type = queue.remove();
+        flatTypes.add(type);
+        queue.addAll(Optional.ofNullable(type.getChildren()).orElse(ImmutableList.of()));
+      }
+      return flatTypes;
+    }
+
+    @Override
+    public TypeDescription record(Types.StructType iStruct, TypeDescription record,
+                                  List<String> names, List<TypeDescription> fields) {
+      return record;
+    }
+
+    @Override
+    public TypeDescription list(Types.ListType iList, TypeDescription array, TypeDescription element) {
+      columnsInContainers.addAll(flatten(element));
+      return null;
+    }
+
+    @Override
+    public TypeDescription map(Types.MapType iMap, TypeDescription map,
+                    TypeDescription key, TypeDescription value) {
+      columnsInContainers.addAll(flatten(key));
+      columnsInContainers.addAll(flatten(value));
+      return null;
```

Comment:
Returning `null` seems odd because it would make a map of maps behave differently. Why not return `map` here instead? The columns will be deduplicated by the set.

Reply:
I had it returning `map` but changed it to `null` to avoid traversals in `flatten` for elements that we'd already have since the top level container elements, even though Set already dedups the added columns. I have no strong opinion on this, I'll change back to map.